apps/src/gamelab/animationListModule.js

```diff
@@ -0,0 +1,605 @@
+
+    case ADD_ANIMATION:
+    case ADD_ANIMATION_AT:
+      return action.data;
+
+    case EDIT_ANIMATION:
+      return Object.assign({}, state, action.data, {
+        saved: false // Dirty, so it'll get saved soon.
+      });
+
+    case INVALIDATE_ANIMATION:
+      return Object.assign({}, state, {
+        saved: false // Dirty, so it'll get saved soon.
+      });
+
+    case SET_ANIMATION_NAME:
+      return Object.assign({}, state, {
+        name: action.name
+      });
+
+    case START_LOADING_FROM_SOURCE:
+      return Object.assign({}, state, {
+        loadedFromSource: false
+      });
+
+    case DONE_LOADING_FROM_SOURCE:
+      return Object.assign({}, state, {
+        loadedFromSource: true,
+        saved: true,
+        blob: action.blob,
+        dataURI: action.dataURI
+      });
+
+    case ON_ANIMATION_SAVED:
+      return Object.assign({}, state, {
+        saved: true,
+        version: action.version
+      });
+
+    default:
+      return state;
+  }
+}
+
+/**
+ *
+ * @param {!SerializedAnimationList} serializedAnimationList
+ * @returns {function()}
+ */
+export function setInitialAnimationList(serializedAnimationList) {
+  return dispatch => {
+    // TODO: Should I validate here?
+    // TODO: Should I add transient properties here?
+    dispatch({
+      type: SET_INITIAL_ANIMATION_LIST,
+      animationList: serializedAnimationList
+    });
+    serializedAnimationList.list.forEach(key => {
+      dispatch(loadAnimationFromSource(key));
+    });
+  };
+}
+
+/**
+ * Add an animation to the project (at the end of the list).
+ * @param {!AnimationKey} key
+ * @param {!SerializedAnimation} data
+ */
+export function addAnimation(key, data) {
+  // TODO: Validate that key is not already in use?
+  // TODO: Validate data format?
+  return dispatch => {
+    dispatch({
+      type: ADD_ANIMATION,
+      key,
+      data
+    });
+    dispatch(loadAnimationFromSource(key, undefined, () => {
+      dispatch(selectAnimation(key));
+    }));
+    dashboard.project.projectChanged();
+  };
+}
+
+/**
+ * Add a library animation to the project.
+ * @param {!SerializedAnimation} data
+ */
+export function addLibraryAnimation(data) {
+  return dispatch => {
+    const key = utils.createUuid();
+    dispatch({
+      type: ADD_ANIMATION,
+      key,
+      data
+    });
+    dispatch(loadAnimationFromSource(key, data.sourceUrl, () => {
+      dispatch(invalidateAnimation(key));
+      dispatch(selectAnimation(key));
+    }));
+    dashboard.project.projectChanged();
+  };
+}
+
+/**
+ * Mark an animation as needing to be saved to S3.
+ * @param {AnimationKey} key
+ * @returns {{type: *, key: *}}
+ */
+function invalidateAnimation(key) {
+  return {
+    type: INVALIDATE_ANIMATION,
+    key
+  };
+}
+
+/**
+ * Clone the requested animation, putting the new one directly after the original
+ * in the project animation list.
+ * @param {!AnimationKey} key
+ * @returns {Function}
+ */
+export function cloneAnimation(key) {
```

Comment:
TODO: `cloneAnimation` shouldn't make an async request itself anymore - it should just copy the locally cached version of another animation, slap a new key on it, and wait for the normal save process to upload it.